### DISCUSSION TITLE
debug mode when linking with static library

### DISCUSSION THREAD

**Discussion by 0wen101** [AmigaPorts/ACE#224] [Q&A]
Hi again, 

Is it possible to put ace into debug mode when compiling and linking with standalone and installed ACE object files ?

ACE helloword for example requires debug mode to write messages to that log file.
If link with the installed ACE I get message to say ACE_DEBUG not used  :

```
D:\Dev\Amiga\ACE\Helloworld\build>cmake .. -DCMAKE_TOOLCHAIN_FILE=C:\AmigaCMakeCrossToolchains\m68k-amigaos.cmake -DTOOLCHAIN_PREFIX=m68k-amigaos -DTOOLCHAIN_PATH=C:\amiga-gcc -DM68K_CPU=68000 -DM68K_FPU=soft -DACE_DEBUG=ON -G "MinGW Makefiles"
M68K_CRT not set, defaulting C Runtime to libnix
-- The C compiler identification is GNU 6.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/amiga-gcc/bin/m68k-amigaos-gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done (1.3s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    ACE_DEBUG

-- Build files have been written to: D:/Dev/Amiga/ACE/Helloworld/build
```
I tried setting ace debug on for the cmake build command to build ACE that way:

`cmake .. -DCMAKE_TOOLCHAIN_FILE=C:\AmigaCMakeCrossToolchains\m68k-amigaos.cmake -DTOOLCHAIN_PREFIX=m68k-amigaos -DTOOLCHAIN_PATH=C:\amiga-gcc -DACE_DEBUG=ON -G "MinGW Makefiles"`

 install again and rebuild helloworld with that version. Same result -  'ACE_DEBUG' not used

Maybe it's the way am linking in cmakelists.txt as when enable :

```
# find_package(ace REQUIRED)
# include_directories(${ace_INCLUDE_DIRS})
```
I get errors re ' Add the installation prefix of "ace" to CMAKE_PREFIX_PATH' etc. 
I'm not sure what 'find_package' is trying to resolve or look for there.

I can only get it to link if add this before last line of cmakelists.txt (ie where my ace headers and objects are installed)

```
include_directories("C:/amiga-gcc/usr/include")
link_directories("C:/amiga-gcc/usr/lib")
```
Resulting size of exe is a lot smaller than the debug version so am assuming it's in release mode.

No issue at all if build helloworld using ace as submodule. Just takes longer to build and get it running that way.

Also, not sure it's possible, but what should/would I add to cmakelists.txt to link with the static library libace.a ? 
I see that's built also along with the object files in that C:/amiga-gcc/usr/lib folder.

**Comment by tehKaiN** [comment]
I  strongly recommend that you keep using submodule builds - when set up properly, ACE should be built only once, and then only files in your project which have been actually changed. Of course there will be ACE rebuilds when you change the debug/release target etc., but I imagine you're using same configuration during 90% of development time.

If that's a no-go for you, `-DACE_DEBUG=ON` for building ACE as well as the project itself should do the trick:

- when building the library, it actually makes it emit required code for using the debug
- when building the game, it actually makes game use debug variants of functions and/or special macros

To make the static link discoverable, you probably need to call `mingw32-make install` after building the libace.a - this will move the library to directory where your Amiga compiler expects to find static libraries. If you don't want to do this, you'll have to add include/link dirs as you've done it already.

it's kinda strange that this define seems to be unused. Do you have this kind of printout? What does it print?

```
[cmake] -- [ACE] ACE_LIBRARY_KIND: 'OBJECT'
[cmake] -- [ACE] ACE_DEBUG: 'OFF'
[cmake] -- [ACE] ACE_DEBUG_UAE: 'OFF'
[cmake] -- [ACE] ACE_BOB_WRAP_Y: 'OFF'
[cmake] -- [ACE] ACE_USE_ECS_FEATURES: 'OFF'
```

**Reply by 0wen101** [reply]
thanks, yes have the -DACE_DEBUG=ON argument for both library and helloworld. cmake command.
I notice larger file sizes for the installed library object files with that argument. Still getting the CMake Warning though 'Manually-specified variables were not used by the project'

Output using sub module:

```
D:\Dev\Amiga\ACE\Helloworld\build>cmake .. -DCMAKE_TOOLCHAIN_FILE=C:\AmigaCMakeCrossToolchains\m68k-amigaos.cmake -DTOOLCHAIN_PREFIX=m68k-amigaos -DTOOLCHAIN_PATH=C:\amiga-gcc -DM68K_CPU=68000 -DM68K_FPU=soft -DACE_DEBUG=ON -G "MinGW Makefiles"
M68K_CRT not set, defaulting C Runtime to libnix
-- The C compiler identification is GNU 6.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/amiga-gcc/bin/m68k-amigaos-gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- [ACE] ACE_LIBRARY_KIND: 'OBJECT'
-- [ACE] ACE_DEBUG: 'ON'
-- [ACE] ACE_DEBUG_UAE: 'OFF'
-- [ACE] ACE_BOB_WRAP_Y: 'ON'
-- [ACE] ACE_USE_ECS_FEATURES: 'OFF'
-- Configuring done (2.5s)
-- Generating done (0.0s)
-- Build files have been written to: D:/Dev/Amiga/ACE/Helloworld/build
```

Output using installed ace library objects:

```
D:\Dev\Amiga\ACE\Helloworld\build>cmake .. -DCMAKE_TOOLCHAIN_FILE=C:\AmigaCMakeCrossToolchains\m68k-amigaos.cmake -DTOOLCHAIN_PREFIX=m68k-amigaos -DTOOLCHAIN_PATH=C:\amiga-gcc -DM68K_CPU=68000 -DM68K_FPU=soft -DACE_DEBUG=ON -G "MinGW Makefiles"
M68K_CRT not set, defaulting C Runtime to libnix
-- The C compiler identification is GNU 6.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/amiga-gcc/bin/m68k-amigaos-gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done (1.3s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    ACE_DEBUG

-- Build files have been written to: D:/Dev/Amiga/ACE/Helloworld/build
```
build the amiga executable:

```
cmake --build .
or
mingw32-make.exe hello
```

Had planned a written tutorial later and a debug static library for helloworld would be good as would save a few steps.

thanks

ps I did also try the argument 
`-DCMAKE_BUILD_TYPE=Debug`
for both library and executable. 
Produces mammoth size files ok but same result as above both in the cmake output and when run the hello executable in winuae. No log file created.

**Reply by tehKaiN** [reply]
okay, I think I know where the problem is, and it's definitely a bug on ACE's side. Here are the details: https://github.com/AmigaPorts/ACE/issues/225 .

I'll try to fix it soon, but my tempo is not that great nowadays, so it could take a week or a month - don't hold your breath on this one. But, for now I'd suggest either using workaround or submodule-based builds. For workaround, look up the ACE's cmakelists, find the code regarding its build switches, e.g.

```cmake
if(ACE_DEBUG)
	target_compile_definitions(${TARGET_NAME} PUBLIC ACE_DEBUG)
endif()
```

and add the relevant line somewhere after your `add_executable` call in your cmakelists file:

```cmake
add_executable(my_awesome_game ...)
target_compile_definitions(my_awesome_game PUBLIC ACE_DEBUG)
```

Re `DCMAKE_BUILD_TYPE=Debug` builds: don't use them unless you use a compiler suite which allows you to attach a debugger and execute your code line by line. This effectively disables all code optimizations/cleanups made by compiler, so the resulting code is humongous and horribly slow. On the other hand, `ACE_DEBUG` just adds logs and runtime checks in most error-prone places, making game a bit slower, but easier to diagnose its misbehavior.

**Reply by 0wen101** [reply]
Ah, thanks will try that with the helloworld cmakelists and see if can turn on the right define to get it into debug mode.
Also if dont mind, am still filling in blanks on the cmake/make file building, defines and build flags etc

Helloworld\CMakeLists.txt has

```
# Debugging flag exposed via CMake specifically for your game code
if(GAME_DEBUG)
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DGAME_DEBUG")
endif()
```

How does the the cmake '-DACE_DEBUG=ON' argument variable enable or set the debug define on in that case? 
Seems like it's expecting GAME_DEBUG instead of ACE_DEBUG

Is 'GAME_DEBUG' something ACE turns on when build the library with submodule (but not with the static lib) ?
If remove the if() and just have : 

`  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DGAME_DEBUG")`

not quite, but I half expected that to do the trick

**Reply by tehKaiN** [reply]
`GAME_DEBUG` is something different than `ACE_DEBUG` - it's a stub in sample cmakelists so that you get the idea on how to enable custom checks for your game logic, as an additional debugging layer. If you look at ACE's cmakelists you'll see that there's `ACE_DEBUG` switch and e.g. ace/managers/memory.h depends on it.

In order for ACE to be run in its debug mode, when built as a standalone library, two things need to happen:
- `ACE_DEBUG` must be set when ACE is built, so that its .c and .h files properly switch things on and the debug functions are actually built,
- `ACE_DEBUG` must be set when the game is built, so that debug-related stuff in ACE's .h files will be used.

**Reply by 0wen101** [reply]
that makes sense and why you'd have to explicitly add another 'target_compile_definitions' to your game cmakelists when using ace as static lib.
Retested helloworld with your fix and works fine now. No more CMake Warning, Log file generates, no further issue.
thank you
